cachito/workers/pkg_managers/npm.py

```diff
@@ -27,35 +29,356 @@
+        For v1 Packages, this will be True if this Package has no dependent Packages
+        """
+        if self.path is None:  # v1 Packages
+            return not bool(self.dependent_packages)
+
+        return Path(self.path).parent.as_posix() in (".", "node_modules")
+
+    def get_dependency_names(self) -> list[str]:
+        """Get the list of names of dependencies that this Package depends on.
+
+        This will be the `dependencies` keys for a given `package` in a
+        v2+ package-lock.json file. It will be the `requires` keys for a given
+        `dependency` in a v1 package-lock.json file.
+        """
+        if self.path is None:  # v1 Packages
+            return list(self._package_dict.get("requires", {}).keys())
+
+        dep_names: set[str] = set()
+        for dep_type in (
+            "dependencies",
+            "devDependencies",
+            "optionalDependencies",
+            "peerDependencies",
+        ):
+            dep_names.update(self._package_dict.get(dep_type, {}).keys())
+
+        return list(dep_names)
+
+    def replace_dependency_version(self, name: str, version: str) -> None:
+        """Replace the version of a dependency of this Package.
+
+        For v1 Packages, this will replace the version of the dependency in the
+        `requires` dict. For v2 Packages it will do the same in the `dependencies` dict.
+        :param name: the name of the dependency that will be updated
+        :param version: the updated version of the dependency
+        """
+        if self.path is None:  # v1 Packages
+            if name in self._package_dict.get("requires", {}):
+                self._package_dict["requires"][name] = version
+        else:
+            for dep_type in (
+                "dependencies",
+                "devDependencies",
+                "optionalDependencies",
+                "peerDependencies",
+            ):
+                if name in self._package_dict.get(dep_type, {}):
+                    self._package_dict[dep_type][name] = version
+
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(other, Package):
+            return (
+                self.name == other.name
+                and self.path == other.path
+                and self._package_dict == other._package_dict
+                and self.alias == other.alias
+                and self.dependent_packages == other.dependent_packages
+            )
+        return False
+
+    def __repr__(self) -> str:
+        attr_str = ", ".join(f"{k}={v}" for k, v in self.__dict__.items())
+        return f"{self.__class__.__name__}({attr_str})"
+
+
+class PackageLock:
+    """A npm package-lock.json file."""
+
+    def __init__(self, lockfile_path: Path, lockfile_data: dict[str, Any]) -> None:
+        """Initialize a PackageLock."""
+        self._lockfile_path = lockfile_path
+        self._lockfile_data = lockfile_data
+        self._original_lockfile_data = copy.deepcopy(lockfile_data)
```

Comment:
I was wondering if the deepcopy here could be expensive, as some lockfiles can be a few megabytes large, but it doesn't seem so.